src/main/java/net/sf/mzmine/modules/peaklistmethods/filtering/duplicatefilter/DuplicateFilterTask.java

```diff
@@ -137,92 +139,153 @@ public void run() {
+          new PeakListRowSorter(SortingProperty.Area, SortingDirection.Descending));
+    else
+      Arrays.sort(peakListRows,
+          new PeakListRowSorter(SortingProperty.ID, SortingDirection.Ascending));
 
     // Loop through all peak list rows
     processedRows = 0;
+    int n = 0;
     totalRows = rowCount;
     for (int firstRowIndex = 0; !isCanceled() && firstRowIndex < rowCount; firstRowIndex++) {
 
-      final PeakListRow firstRow = peakListRows[firstRowIndex];
-      if (firstRow != null) {
+      final PeakListRow mainRow = peakListRows[firstRowIndex];
+      if (mainRow != null) {
+        // need to copy to not alter the old peakList
+        PeakListRow firstRow = copyRow(mainRow);
 
         for (int secondRowIndex = firstRowIndex + 1; !isCanceled()
             && secondRowIndex < rowCount; secondRowIndex++) {
 
           final PeakListRow secondRow = peakListRows[secondRowIndex];
           if (secondRow != null) {
-
             // Compare identifications
             final boolean sameID =
                 !requireSameId || PeakUtils.compareIdentities(firstRow, secondRow);
 
-            // Compare m/z
-            final boolean sameMZ = mzTolerance.getToleranceRange(firstRow.getAverageMZ())
-                .contains(secondRow.getAverageMZ());
-
-            // Compare rt
-            final boolean sameRT = rtTolerance.getToleranceRange(firstRow.getAverageRT())
-                .contains(secondRow.getAverageRT());
+            boolean sameMZRT = false;
+            if (!mode.equals(FilterMode.SINGLE_FEATURE)) {
+              // Compare m/z
+              boolean sameMZ = mzTolerance.getToleranceRange(firstRow.getAverageMZ())
+                  .contains(secondRow.getAverageMZ());
+
+              // Compare rt
+              boolean sameRT = rtTolerance.getToleranceRange(firstRow.getAverageRT())
+                  .contains(secondRow.getAverageRT());
+              sameMZRT = sameMZ && sameRT;
+            } else {
+              // at least one similar feature in one raw data file
+              for (RawDataFile raw : rawFiles) {
+                Feature f1 = firstRow.getPeak(raw);
+                Feature f2 = secondRow.getPeak(raw);
+                if (f1 != null && f2 != null) {
+                  // Compare m/z
+                  boolean sameMZ = mzTolerance.getToleranceRange(firstRow.getAverageMZ())
+                      .contains(secondRow.getAverageMZ());
+
+                  // Compare rt
+                  boolean sameRT = rtTolerance.getToleranceRange(firstRow.getAverageRT())
+                      .contains(secondRow.getAverageRT());
+                  sameMZRT = sameMZ && sameRT;
+
+                  // at least one
+                  if (sameMZRT)
+                    break;
+                }
+              }
+            }
 
             // Duplicate peaks?
-            if (sameID && sameMZ && sameRT) {
-
-              peakListRows[secondRowIndex] = null;
+            if (sameID && sameMZRT) {

```

Comment:
It's hard to understand these nested loops and if statements. Any way to turn them into separate functions with self-explanatory names?